tests/rptest/tests/compatibility/kafka_streams_test.py

```diff
@@ -261,32 +288,68 @@ class KafkaStreamsAnomalyDetection(KafkaStreamsProdConsBase):
         TopicSpec(name="AnomalousUsers"),
     )
 
-    Example = KafkaStreamExamples.KafkaStreamsAnomalyDetection
-    PRODUCER = AnomalyProducer
-
     def __init__(self, test_context):
         super(KafkaStreamsAnomalyDetection,
-              self).__init__(test_context=test_context,
-                             enable_pp=True,
-                             enable_sr=True)
+              self).__init__(test_context=test_context)
+        self._ctx = test_context
+        self._num_records = 10 if self.scale.local else 1000
 
     def is_valid_msg(self, msg):
-        key = msg["key"]
-        return "record-1" in key or "record-2" in key or "record-3" in key
+        username = msg["key"]
+        count = int(msg["value"])
+
+        if "user-1" in username:
+            return 2 < count and count <= (self._num_records * 2)
+        else:
+            return 2 < count and count <= self._num_records
+
+    @cluster(num_nodes=6)
+    def test_kafka_streams_anomaly_detection(self):
+        example_jar = KafkaStreamExamples.AnomalyDetectionExample(
+            self.redpanda)
+        example = ExampleRunner(self._ctx, example_jar, timeout_sec=TIMEOUT)
+
+        producer = AnomalyProducer(self._ctx, self.redpanda,
+                                   self.topics[0].name, self._num_records)
+
+        form = '\'%{"key":"%k", "value":"%v{unpack[>Q]}"%}\\n\''
+        consumer = RpkConsumer(self._ctx,
+                               self.redpanda,
+                               self.topics[1].name,
+                               formatter=form)
+
+        # Start the example
+        example.start()
+        example.wait()
+
+        # Produce some data
+        producer.start()
+        producer.wait()
+
+        # Consume the data
+        consumer.start()
+        # is_valid_msg is called on every message the consumer reads
+        consume(self._ctx, consumer, self.is_valid_msg)
```

Comment:
It seems there may be an imprecision. The test has a straightforward shape:

 - do an action
 - consume the messages and check that they are valid

What concerns me is that `consume` doesn't check all messages but only a non deterministic number of messages.
For example if a test wrote 5 messages `consume` may check only the first two.

Do you know if it's possible the amount of messages we need to read ahead of time or maybe the last message is special and the consumer may read until it meets it.

Reply:
I see your point. Yes, we know the number of messages ahead of time and I can use that number on the consume loop.
I'll make the change.

Reply:
Actually, I use RpkConsumer in these tests which by default starts consume from the oldest offset. Therefore, all messages from the topic are fetched. But I still agree with you that explicitly checking for the number of messages is better.